frontend/dockerfile/dockerfile2llb/convert.go

```diff
@@ -756,9 +832,15 @@ func dispatch(d *dispatchState, cmd command, opt dispatchOpt) error {
 			location:     c.Location(),
 			opt:          opt,
 		})
-		if err == nil && len(cmd.sources) == 0 {
-			for _, src := range c.SourcePaths {
-				d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
+		if err == nil {
+			if len(cmd.sources) == 0 {
+				for _, src := range c.SourcePaths {
+					d.ctxPaths[path.Join("/", filepath.ToSlash(src))] = struct{}{}
+				}
+			} else if source := cmd.sources[0]; source.namedCtxState != nil {
+				for _, src := range c.SourcePaths {
+					source.namedCtxState.includePath(src)
+				}
```

Comment:
I think this logic would need to be replicated to the `instructions.AddCommand` branch as well.

Also, ideally we could somehow just have a single combined struct that wraps both a stage source, and a named context.

Reply:
I don't think this needs to be replicated to `ADD`. `ADD` doesn't seem to support the `--from` option. The `CopyCommand` has a `From` option while `AddCommand` doesn't seem to have that.